Use smaller cursor sizes

stefan11111 opened the pull request:
Split from https://github.com/X11Libre/xserver/pull/734

Still incomplete, as on most hw this just uses the largest cursor available, and coloring the cursor buffer is really inefficient, but I have to pick somewhere to split the above pr.

Part-of: #549
Part-of: #675

stefan11111 commented:
@notbabaisyou Ping, as I couldn't add you as a reviewer through the github interface.

metux commented:
@stefan11111 theoretically you should also be able to assign @X11Libre/wranglers as reviewers.
Semantically, @X11Libre/reviewers would be more correct, but I don't know why this isn't offered as reviewers yet.
@callmetango any idea whether we need some github config tweak for that ?

stefan11111 commented:
> Picked the first two commits, for the remaining like to hear some more other oppinions. (for strange reasons github doesn't show that the first two already in master)

Rebased now.

stefan11111 commented:
> Do we ever want to use a larger cursor that this?
> 
> This function determines the smallest cursor size that is at least as big as the smallest size supported by all crtc's.
> 
> As I understand it, after we pass this size to xf86_cursors_init, the cursor image will have this size, and the cursor glyph will be at most this size.
> 
> When we then set the hw cursor size, we always pick the smallest one, that is at least as big as the cursor glyph, and I also added checks to make sure that we never go over this size.

yes large cursors are great for presentations or being old as dirt with bad eyes

_Originally posted by @HaplessIdiot in https://github.com/X11Libre/xserver/pull/734#discussion_r2309906292_

stefan11111 commented:
> > Do we ever want to use a larger cursor that this?
> > This function determines the smallest cursor size that is at least as big as the smallest size supported by all crtc's.
> > As I understand it, after we pass this size to xf86_cursors_init, the cursor image will have this size, and the cursor glyph will be at most this size.
> > When we then set the hw cursor size, we always pick the smallest one, that is at least as big as the cursor glyph, and I also added checks to make sure that we never go over this size.
> 
> yes large cursors are great for presentations or being old as dirt with bad eyes
> 
> _Originally posted by @HaplessIdiot in [#734 (comment)](https://github.com/X11Libre/xserver/pull/734#discussion_r2309906292)_

As this pr is written right now, the cursor buffer is at least 64x64, which is quite big already.
To see how big that is, change the paint cursor function from this pr to paint the entire cursor buffer black (memset to -1) and send the width and height to 64x64.

Also, in the bigger pr I also add an option for setting the cursor to be as big as desired, so this will be adjustable in xorg.conf.

Also, the cursor buffer size is just an upper bound for the cursor glyph. If the glyph is 9 x 16, like it is on my machine usually, the rendered cursor will look that same regardless of if the cursor buffer is 32x32, 64x64 or 256x256.

notbabaisyou commented on hw/xfree86/drivers/video/modesetting/driver.c:
```diff
@@ -1953,14 +1953,9 @@ ScreenInit(ScreenPtr pScreen, int argc, char **argv)
     /* Need to extend HWcursor support to handle mask interleave */
     if (!ms->drmmode.sw_cursor)
-        xf86_cursors_init(pScreen, cursor_dim.width, cursor_dim.height,
+        xf86_cursors_init(pScreen, ms->cursor_image_width, ms->cursor_image_height,
```
Is using a single size for everything really a good idea? This'll cause issues for anyone using a larger than the smallest functioning cursor.

stefan11111 replied:
It was already a single size for everything in the old code.

notbabaisyou replied:
The old code used `min_cursor_[width|height]`  for the minimum cursor size which it got through probing and `max_cursor_[width|height]` was the maximum cursor size which was given through `DRM_CAP_CURSOR_[WIDTH|HEIGHT]`

notbabaisyou commented on hw/xfree86/drivers/video/modesetting/drmmode_display.c:
```diff
@@ -1823,27 +1823,65 @@ drmmode_set_cursor(xf86CrtcPtr crtc, int width, int height)
+    int width  = drmmode_cursor->dimensions[idx].width;
+    int height = drmmode_cursor->dimensions[idx].height;
+
+    int num_pitches = drmmode_cursor->num_dimensions;
```
Shouldn't we simply be checking if `num_dimensions` is equal to one and then doing this?

stefan11111 replied:
We're only probing it once, so there isn't any noticeable overhead here.
After probing the pitch, we cache it and return the cached pitch.

notbabaisyou commented on hw/xfree86/drivers/video/modesetting/drmmode_display.c:
```diff
@@ -1855,43 +1893,46 @@ static Bool
+        if (dimensions.width > max_width ||
+            dimensions.height > max_height) {
+
+            /* If this ever happens, i should not be 0, but check for good measure */
```
I'm still against this check since it's completely redundant.

stefan11111 replied:
It is redundant, but it's just intended as a sanity check.
I can remove it if desired.

notbabaisyou commented on hw/xfree86/drivers/video/modesetting/drmmode_display.c:
```diff
@@ -4591,20 +4635,12 @@ drmmode_create_initial_bos(ScrnInfoPtr pScrn, drmmode_ptr drmmode)
 
     bpp = 32;
 
+    drmmode_get_smallest_supported_cursor(pScrn, &ms->cursor_image_width, &ms->cursor_image_height);
```
Using the smallest cursor available will break hardware cursors the moment you go above the smallest size as explained above. My PR did exactly what was done before and that is use the largest cursor available for the buffer but tell the hardware to only use a part of it.

stefan11111 replied:
The way the hw cursor code is designed, the cursor glyph will never be bigger than the size we tell it to use as the max, so we should never go over this size when setting the actual hw cursor size with ioctl's.

notbabaisyou commented on hw/xfree86/drivers/video/modesetting/drmmode_display.c:
```diff
@@ -4741,37 +4777,31 @@ drmmode_crtc_set_vrr(xf86CrtcPtr crtc, Bool enabled)
-            max_width  = width;
-            max_height = height;
-        }
+        max_size.width = (width > max_size.width) ? width : max_size.width;
```
Not a fan of in-lining these if-else statements.

stefan11111 replied:
The inlined statements are not equivalent to the above code.
The old code increased the width and height at the same time, and only if both were bigger that the old maximums.
This code increases the minimums separately.

For example, if one crtc supported 64x128 as it's minimum, and another supported 128x64, the old code would pick the size as 64x128 instead on 128x128.

notbabaisyou commented on hw/xfree86/drivers/video/modesetting/drmmode_display.h:
```diff
@@ -245,6 +245,8 @@ typedef struct {
     Bool vrr_enabled;
     Bool use_gamma_lut;
+
+    int* cursor_pitches;
```
I think you are missing a call to `free()`.

stefan11111 replied:
You're right, I should probably free it in drmmode_crtc_destroy.

Btw, in drmmode_crtc_destroy, we do this:
```
    if (!ms->atomic_modeset)
        return;

    drmmode_prop_info_free(drmmode_crtc->props_plane, DRMMODE_PLANE__COUNT);
    free(drmmode_crtc->cursor.dimensions);
```
Shouldn't we free the cursor dimensions regardless of whether or not we use atomic modesetting?
I'm thinking we should do the same for pitches here, like this:
```
free(drmmode_crtc->cursor_pitches); /* and maybe NULL for good measure? */
free(drmmode_crtc->cursor.dimensions);

    if (!ms->atomic_modeset)
        return;
```
What do you think?

notbabaisyou replied:
Fine by me, I didn't notice that it would only free in an atomic context only since my patches were written on a Xorg DDX that is atomic only.

This is probably worth fixing ASAP since we're leaking memory.

stefan11111 replied:
Done.

notbabaisyou commented on hw/xfree86/drivers/video/modesetting/drmmode_display.c:
```diff
@@ -4741,37 +4777,31 @@ drmmode_crtc_set_vrr(xf86CrtcPtr crtc, Bool enabled)
-    if (!cursor_lim)
-        return FALSE;
+    /* We need the cursor image to be at least 64x64 */
+    drmmode_cursor_dim_rec max_size = {.width = 64, .height = 64};
```
This is redundant since we will already have one known dimension by default that is 64x64. (or whatever the driver provides)

stefan11111 replied:
If the driver gives a size less than 64x64, then this would break without the 64x64 minimum here.

Also, I plan to add a config option to set cursor size in xorg.conf, so this code will be needed so that we don't allocate a buffer smaller that 64x64.

notbabaisyou replied:
I've looked through the kernel drivers and there seem to be no drivers that advertise less than `64x64` for cursors.

If you want to prevent cursor sizes below `64x64` you probably should do that earlier on.

stefan11111 replied:
Where?

notbabaisyou commented on hw/xfree86/drivers/video/modesetting/drmmode_display.c:
```diff
@@ -1855,43 +1893,46 @@ static Bool
 
-        if (dimensions.width > cursor->bits->width &&
-            dimensions.height > cursor->bits->height)
+        if (dimensions.width >= cursor->bits->width &&
```
While we don't have any NPOT (non power of two) cursors to deal with, this logic would probably break with these being combined.

stefan11111 replied:
What? cursor->bits represents the size of the cursor glyph. This check makes it so that the cursor size chosen is at lease equal to the glyph size, while the older code chose a strictly greater size.
I don't see what powers of 2 have to do here?

stefan11111 replied:
> My bad, misunderstood the change.
> 
> I believe this was discussed in a previous change but I still believe we should check if either of dimensions is out of spec instead of if both are out of spec.

You mean that, if the glyph is 17x9, for example, and we have a 16x16 and a 32x32 cursor, we should clip the glyph to 16x9 and use the 16x16 cursor?

notbabaisyou replied:
I was more worried that a non-square cursor would be able to get away within an incompatible dimension. But that probably won't ever be an issue or be the case.

Nevermind.

notbabaisyou reviewed (approved):
I don't like this approach since this just thrashes the kernel with probe attempts, there really isn't a good way to figure out an optimal cursor without the drivers actually telling us.

Aside from that, a bunch of nits but I won't stop you from doing this.

stefan11111 commented on hw/xfree86/drivers/video/modesetting/drmmode_display.c:
```diff
@@ -1823,27 +1823,65 @@ drmmode_set_cursor(xf86CrtcPtr crtc, int width, int height)
+                     const CARD32 * restrict image, int image_width, int image_height)
 {
-    return cursor.dimensions[cursor.num_dimensions - 1].height;
+    if (cursor_width == image_width && cursor_pitch == cursor_width) {
```
memcpy is more concise and imo easier to reason about that those loops.

Btw, if you look at the other pr with more changes, you'll see that I plan to implement some damage-like tracking for the cursor buffer, so we only paint what's changed and not the entire buffer.

metux commented:
once all conversations are finished up and getting some more acks, I'm going to merge this :)

great work, thanks.

stefan11111 commented:
@notbabaisyou @metux @callmetango 
Split the second commit into https://github.com/X11Libre/xserver/pull/983

stefan11111 commented:
@notbabaisyou @b-aaz
I split the pitch commit into https://github.com/X11Libre/xserver/pull/983
Could you comment there about pitch-related changes instead of here.

stefan11111 commented:
Split part of this into https://github.com/X11Libre/xserver/pull/1041

HaplessIdiot commented:
This is a very important change if we arent using the full cursor buffer why waste latency running the whole buffer this is excellent!

stefan11111 commented on hw/xfree86/drivers/video/modesetting/drmmode_display.c:
```diff
@@ -1825,10 +1826,10 @@ drmmode_set_cursor(xf86CrtcPtr crtc, int width, int height)
     }
 
-    if (ret)
+    if (ret) {
```
done

stefan11111 commented:
Rebased and simplified following https://github.com/X11Libre/xserver/pull/1109

@metux @callmetango @b-aaz @cepelinas9000 @algrid @X11Libre/reviewers ping

metux commented:
Looks good to me, but maybe somebody with more actual HW can test it ?
@josephcrowell @cepelinas9000
